File: src/generate/surge.rs
use crate::config::AppConfig;
use crate::model::ProxyNode;

/// Generate Surge [Proxy] lines for a list of nodes.
pub fn generate_surge_proxy_list(nodes: &[ProxyNode], listen_addr: &str) -> String {
    nodes
        .iter()
        .map(|n| format!("{} = socks5, {}, {}", n.name, listen_addr, n.assigned_port))
        .collect::<Vec<_>>()
        .join("\n")
}

/// Generate a full Surge config snippet with [Proxy] and [Proxy Group] sections.
pub fn generate_surge_config(
    config: &AppConfig,
    groups: &indexmap::IndexMap<String, Vec<ProxyNode>>,
) -> String {
    let listen_addr = &config.port.listen_addr;
    let server_listen = &config.server.listen;

    let mut lines = Vec::new();
    lines.push("[Proxy]".to_string());

    for nodes in groups.values() {
        for n in nodes {
            lines.push(format!(
                "{} = socks5, {}, {}",
                n.name, listen_addr, n.assigned_port
            ));
        }
    }

    lines.push(String::new());
    lines.push("[Proxy Group]".to_string());

    for (name, _) in groups {
        lines.push(format!(
            "{name} = select, policy-path=http://{server_listen}/surge/group/{name}"
        ));
    }

    // AllNodes group referencing all sub-groups
    if groups.len() > 1 {
        let group_names: Vec<_> = groups.keys().cloned().collect();
        lines.push(format!("AllNodes = select, {}", group_names.join(", ")));
    }

    lines.join("\n")
}

File: zig/src/generate/surge.zig
const std = @import("std");
const config = @import("../config.zig");
const model = @import("../model.zig");

pub fn generateSurgeProxyList(gpa: std.mem.Allocator, nodes: []const model.ProxyNode, listen_addr: []const u8) ![]const u8 {
    var lines = std.ArrayList(u8).empty;
    defer lines.deinit(gpa);
    try lines.ensureTotalCapacity(gpa, nodes.len * 64);
    for (nodes) |n| {
        try lines.print(gpa, "{s} = socks5, {s}, {d}\n", .{ n.name, listen_addr, n.assigned_port });
    }
    return lines.toOwnedSlice(gpa);
}

pub fn generateSurgeConfig(
    gpa: std.mem.Allocator,
    cfg: config.AppConfig,
    groups: *const std.array_hash_map.String([]model.ProxyNode),
) ![]const u8 {
    const listen_addr = cfg.port.listen_addr;
    const server_listen = cfg.server.listen;

    var lines = std.ArrayList(u8).empty;
    defer lines.deinit(gpa);
    var total_nodes: usize = 0;
    var cit = groups.iterator();
    while (cit.next()) |entry| total_nodes += entry.value_ptr.*.len;
    try lines.ensureTotalCapacity(gpa, total_nodes * 64 + 256);

    try lines.appendSlice(gpa, "[Proxy]\n");
    var git = groups.iterator();
    while (git.next()) |entry| {
        for (entry.value_ptr.*) |n| {
            try lines.print(gpa, "{s} = socks5, {s}, {d}\n", .{ n.name, listen_addr, n.assigned_port });
        }
    }

    try lines.appendSlice(gpa, "\n[Proxy Group]\n");
    git = groups.iterator();
    while (git.next()) |entry| {
        const name = entry.key_ptr.*;
        try lines.print(gpa, "{s} = select, policy-path=http://{s}/surge/group/{s}\n", .{ name, server_listen, name });
    }

    if (groups.count() > 1) {
        try lines.appendSlice(gpa, "AllNodes = select, ");
        var first = true;
        git = groups.iterator();
        while (git.next()) |entry| {
            if (!first) try lines.appendSlice(gpa, ", ");
            first = false;
            try lines.appendSlice(gpa, entry.key_ptr.*);
        }
        try lines.appendSlice(gpa, "\n");
    }

    return lines.toOwnedSlice(gpa);
}
